File: mcp-server/typescript/src/audit.ts
/**
 * Structured audit logging for every tool invocation (stderr only — stdout
 * is reserved for the MCP JSON-RPC frames).
 *
 * Implements §10 of ALGOVOI_MCP.md.  Raw arguments are hashed, never logged.
 */

import { createHash, randomBytes } from "node:crypto";

function hashArgs(args: unknown): string {
  let payload: string;
  try {
    payload = JSON.stringify(args, Object.keys(args ?? {}).sort());
  } catch {
    payload = String(args);
  }
  return createHash("sha256").update(payload).digest("hex").slice(0, 16);
}

export interface AuditEntry {
  tool_name: string;
  args: unknown;
  status: "ok" | "error" | "rejected";
  duration_ms: number;
  error_code?: string;
}

export function logCall(entry: AuditEntry): void {
  const record: Record<string, unknown> = {
    timestamp:   new Date().toISOString(),
    trace_id:    randomBytes(8).toString("hex"),
    tool_name:   entry.tool_name,
    args_hash:   hashArgs(entry.args),
    status:      entry.status,
    duration_ms: Math.round(entry.duration_ms * 100) / 100,
  };
  if (entry.error_code) record.error_code = entry.error_code;
  try {
    process.stderr.write(JSON.stringify(record) + "\n");
  } catch {
    // Never let audit log failures break a tool call.
  }
}

File: mcp-server/python/algovoi_mcp/audit.py
"""
Structured audit logging for every tool invocation.

Implements §10 of ALGOVOI_MCP.md — each call emits a single-line JSON
record to **stderr** (stdout is reserved for the MCP protocol frames).
Raw arguments are hashed, not logged, so secrets that somehow leak into
the argument map (e.g. a webhook_secret passed explicitly) never hit the
log file.

Format::

    {
      "timestamp":   "2026-04-16T18:04:33Z",
      "trace_id":    "a1b2c3d4e5f6",
      "tool_name":   "create_payment_link",
      "args_hash":   "sha256 prefix",
      "status":      "ok" | "error" | "rejected",
      "duration_ms": 42.18,
      "error_code":  "ValidationError"     # only on non-ok
    }
"""

from __future__ import annotations

import hashlib
import json
import secrets
import sys
import time
from typing import Any


def _hash_args(args: Any) -> str:
    """Stable, redacted 16-char SHA-256 prefix of the arguments.

    We serialise with ``sort_keys=True`` so logically identical calls hash
    to the same prefix regardless of dict ordering, and fall back to
    ``str()`` if some value is not JSON-serialisable.
    """
    try:
        payload = json.dumps(args, sort_keys=True, default=str)
    except Exception:
        payload = repr(args)
    return hashlib.sha256(payload.encode()).hexdigest()[:16]


def log_call(
    *,
    tool_name:   str,
    args:        Any,
    status:      str,
    duration_ms: float,
    error_code:  str | None = None,
) -> None:
    """
    Emit one audit record to ``sys.stderr``.

    Args:
        tool_name:   MCP tool name, e.g. "create_payment_link".
        args:        The raw argument dict (only the SHA-256 prefix is logged).
        status:      One of ``"ok"``, ``"error"``, ``"rejected"``.
        duration_ms: Wall-clock cost of the tool handler.
        error_code:  Optional short error class name (e.g. ``"ValidationError"``).
    """
    entry = {
        "timestamp":   time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
        "trace_id":    secrets.token_hex(8),
        "tool_name":   tool_name,
        "args_hash":   _hash_args(args),
        "status":      status,
        "duration_ms": round(float(duration_ms), 2),
    }
    if error_code:
        entry["error_code"] = error_code
    try:
        sys.stderr.write(json.dumps(entry, separators=(",", ":")) + "\n")
        sys.stderr.flush()
    except Exception:
        # Never let audit-log failures block or fail a tool call.
        pass
